File: src/boat/util/GlHelper.java
package boat.util;

import boat.block.BlockState;
import boat.world.Chunk;
import boat.world.Chunk.Face;
import boat.world.World;

import java.nio.IntBuffer;
import java.util.Random;

import static org.lwjgl.opengl.GL46C.*;

public class GlHelper {
    public static int createShader(int type, String source) {
        int shader = glCreateShader(type);
        glShaderSource(shader, source);
        glCompileShader(shader);
        int[] status = new int[1];
        glGetShaderiv(shader, GL_COMPILE_STATUS, status);
        if (status[0] == GL_TRUE) return shader;
        throw new RuntimeException("Failed to compile " + (type == GL_FRAGMENT_SHADER ? "fragment" : "vertex") + " shader");
    }

    public static int createProgram(int vertShader, int fragShader) {
        int program = glCreateProgram();
        glAttachShader(program, vertShader);
        glAttachShader(program, fragShader);
        glLinkProgram(program);
        int[] status = new int[1];
        glGetProgramiv(program, GL_LINK_STATUS, status);
        if (status[0] == GL_TRUE) return program;
        throw new RuntimeException("Failed to link program");
    }

    public static int dir(Face face, byte x, byte y, byte z) {
        switch (face) {
            case SOUTH -> {
                return z <= 0  ?
                       ((x << 8) | (y << 4) | ((15) << 0)) + 8192:
                       (x << 8) | (y << 4) | ((z - 1) << 0);
            }
            case NORTH -> {
                return z >= 15 ?
                       ((x << 8) | (y << 4) | ((0) << 0)) + 8192:
                       (x << 8) | (y << 4) | ((z + 1) << 0);
            }
            case WEST -> {
                return x <= 0  ?
                       (((15) << 8) | (y << 4) | (z << 0)) + 8192:
                       ((x - 1) << 8) | (y << 4) | (z << 0);
            }
            case EAST -> {
                return x >= 15 ?
                       (((0) << 8) | (y << 4) | (z << 0)) + 8192:
                       ((x + 1) << 8) | (y << 4) | (z << 0);
            }
            case DOWN -> {
                return y <= 0  ?
                       ((x << 8) | ((15) << 4) | (z << 0)) + 8192:
                       (x << 8) | ((y - 1) << 4) | (z << 0);
            }
            case UP -> {
                return y >= 15 ?
                       ((x << 8) | ((0) << 4) | (z << 0)) + 8192:
                       (x << 8) | ((y + 1) << 4) | (z << 0);
            }
        }
        throw new RuntimeException();
    }

    public static void sortBuffer(IntBuffer buffer, int[] array, int[] counts, int startIndex, int endIndex) {
        for (int i = 0, sum = startIndex; i < counts.length; i++) {
            int prev = counts[i];
            counts[i] = sum;
            sum += prev;
        }
        int pos = startIndex;
        while (pos < endIndex) {
            int n = array[pos];
            int v = buffer.get(pos);
            int dest = counts[n & 0xf];

            if (dest <= pos) {
                if (dest < pos) {
                    buffer.put(pos, buffer.get(dest));
                    buffer.put(dest, v);

                    array[pos] = array[dest];
                    array[dest] = n;
                }
                else ++pos;
                counts[n & 0xf] = dest + 1;
            }
            else ++pos;
        }

    }
}

File: src/boat/util/GlHelper.kt
package boat.util

import boat.world.Chunk.Face
import org.lwjgl.opengl.GL11C
import org.lwjgl.opengl.GL20C
import java.nio.IntBuffer

fun createShader(type: Int, source: String?): Int {
    val shader = GL20C.glCreateShader(type)
    GL20C.glShaderSource(shader, source)
    GL20C.glCompileShader(shader)
    val status = IntArray(1)
    GL20C.glGetShaderiv(shader, GL20C.GL_COMPILE_STATUS, status)
    if (status[0] == GL11C.GL_TRUE) return shader
    throw RuntimeException("Failed to compile " + (if (type == GL20C.GL_FRAGMENT_SHADER) "fragment" else "vertex") + " shader")
}

fun createProgram(vertShader: Int, fragShader: Int): Int {
    val program = GL20C.glCreateProgram()
    GL20C.glAttachShader(program, vertShader)
    GL20C.glAttachShader(program, fragShader)
    GL20C.glLinkProgram(program)
    val status = IntArray(1)
    GL20C.glGetProgramiv(program, GL20C.GL_LINK_STATUS, status)
    if (status[0] == GL11C.GL_TRUE) return program
    throw RuntimeException("Failed to link program")
}

fun sortBuffer(buffer: IntBuffer, array: IntArray, counts: IntArray, startIndex: Int, endIndex: Int) {
    var i = 0
    var sum = startIndex
    while (i < counts.size) {
        val prev = counts[i]
        counts[i] = sum
        sum += prev
        i++
    }
    var pos = startIndex
    while (pos < endIndex) {
        val n = array[pos]
        val v = buffer[pos]
        val dest = counts[n and 0xf]
        if (dest <= pos) {
            if (dest < pos) {
                buffer.put(pos, buffer[dest])
                buffer.put(dest, v)
                array[pos] = array[dest]
                array[dest] = n
            } else ++pos
            counts[n and 0xf] = dest + 1
        } else ++pos
    }
}

fun dir(face: Face?, x: Byte, y: Byte, z: Byte): Int {
    when (face) {
        Face.SOUTH -> {
            return if (z <= 0) (x.toInt() shl 8 or (y.toInt() shl 4) or (15 shl 0)) + 8192 else x.toInt() shl 8 or (y.toInt() shl 4) or (z - 1 shl 0)
        }

        Face.NORTH -> {
            return if (z >= 15) (x.toInt() shl 8 or (y.toInt() shl 4) or (0 shl 0)) + 8192 else x.toInt() shl 8 or (y.toInt() shl 4) or (z + 1 shl 0)
        }

        Face.WEST -> {
            return if (x <= 0) (15 shl 8 or (y.toInt() shl 4) or (z.toInt() shl 0)) + 8192 else x - 1 shl 8 or (y.toInt() shl 4) or (z.toInt() shl 0)
        }

        Face.EAST -> {
            return if (x >= 15) (0 shl 8 or (y.toInt() shl 4) or (z.toInt() shl 0)) + 8192 else x + 1 shl 8 or (y.toInt() shl 4) or (z.toInt() shl 0)
        }

        Face.DOWN -> {
            return if (y <= 0) (x.toInt() shl 8 or (15 shl 4) or (z.toInt() shl 0)) + 8192 else x.toInt() shl 8 or (y - 1 shl 4) or (z.toInt() shl 0)
        }

        Face.UP -> {
            return if (y >= 15) (x.toInt() shl 8 or (0 shl 4) or (z.toInt() shl 0)) + 8192 else x.toInt() shl 8 or (y + 1 shl 4) or (z.toInt() shl 0)
        }

        else -> { throw RuntimeException() }
    }
}